src/enclave/httpsig.h

```diff
@@ -0,0 +1,314 @@
+    // ["foo", "bar", "baz"])
+    std::vector<std::string_view> parse_delimited_string(
+      std::string_view& s, const std::string& delimiter)
+    {
+      std::vector<std::string_view> strings;
+      bool last_string;
+
+      auto next_delimiter = s.find(delimiter);
+      while (next_delimiter != std::string::npos || !last_string)
+      {
+        auto token = s.substr(0, next_delimiter);
+        if (next_delimiter == std::string::npos)
+        {
+          last_string = true;
+        }
+
+        strings.emplace_back(token);
+
+        if (!last_string)
+        {
+          s = s.substr(next_delimiter + 1);
+          next_delimiter = s.find(delimiter);
+        }
+      }
+
+      return strings;
+    }
+
+    std::optional<SignatureParams> parse_signature_params(
+      std::string_view& auth_header_value)
+    {
+      SignatureParams sig_params = {};
+
+      auto next_comma = auth_header_value.find(SIGN_PARAMS_DELIMITER);
+      bool last_key = false;
+
+      auto parsed_params =
+        parse_delimited_string(auth_header_value, SIGN_PARAMS_DELIMITER);
+
+      for (auto& p : parsed_params)
+      {
+        LOG_FAIL_FMT(p);
+
+        auto eq_pos = p.find("=");
+        if (eq_pos != std::string::npos)
+        {
+          auto k = p.substr(0, eq_pos);
+          auto v = p.substr(eq_pos + 1);
+
+          // Remove inverted commas around value
+          v.remove_prefix(v.find_first_of("\"") + 1);
+          v.remove_suffix(v.size() - v.find_last_of("\""));
+
+          if (k == SIGN_PARAMS_KEYID)
+          {
+            // keyId is ignored
+          }
+          else if (k == SIGN_PARAMS_ALGORITHM)
+          {
+            sig_params.signature_algorithm = v;
+            if (v != SIGN_ALGORITHM)
+            {
+              LOG_FAIL_FMT("Signature algorithm {} is not supported", v);
+              return {};
+            }
+          }
+          else if (k == SIGN_PARAMS_SIGNATURE)
+          {
+            sig_params.signature = v;
+          }
+          else if (k == SIGN_PARAMS_HEADERS)
+          {
+            auto parsed_signed_headers =
+              parse_delimited_string(v, SIGN_PARAMS_HEADERS_DELIMITER);
+            for (const auto& h : parsed_signed_headers)
+            {
+              LOG_FAIL_FMT(h);
+              sig_params.signed_headers.emplace_back(h);
+            }
+          }
+        }
+        else
+        {
+          LOG_FAIL_FMT("Authorization parameter {} does not contain \"=\"", p);
+          return {};
+        }
+      }
+
+      return sig_params;
+    }
+
+    std::optional<std::vector<uint8_t>> construct_raw_signed_string(
+      const std::vector<std::string_view>& signed_headers)
+    {
+      std::string signed_string = {};
+      for (auto& f : signed_headers)
+      {
+        // Signed headers are listed lowercase in Authorization headers
+        // Uppercase first letter to find corresponding HTTP header
```

Comment:
This doesn't seem like it works in the general case. We may get a header like `Content-Type`, this won't match it. HTTP headers should be treated as case insensitive anyway, so I guess when we construct our `HeaderMap` we should be lowercasing everything?